app/services/fact_importer.rb

```diff
@@ -40,26 +40,58 @@ def initialize(host, facts = {})
   # expect a facts hash
   def import!
     ActiveRecord::Base.transaction do
+      ensure_fact_names
       delete_removed_facts
       update_facts
       add_new_facts
     end
 
-    if @error
-      Foreman::Logging.exception("Error during fact import for #{@host.name}", @error)
-      raise ::Foreman::WrappedException.new(@error, N_("Import of facts failed for host %s"), @host.name)
-    end
+    report_error(@error) if @error
     logger.info("Import facts for '#{host}' completed. Added: #{counters[:added]}, Updated: #{counters[:updated]}, Deleted #{counters[:deleted]} facts")
   end
 
+  def report_error(error)
+    Foreman::Logging.exception("Error during fact import for #{@host.name}", error)
+    raise ::Foreman::WrappedException.new(error, N_("Import of facts failed for host %s"), @host.name)
+  end
+
   # to be defined in children
   def fact_name_class
     raise NotImplementedError
   end
 
   private
 
-  attr_reader :host, :facts
+  attr_reader :host, :facts, :fact_names
+
+  def ensure_fact_names
+    initialize_fact_names
+
+    missing_keys = facts.keys - fact_names.keys
+    add_missing_fact_names(missing_keys)
+  end
+
+  def add_missing_fact_names(missing_keys)
+    missing_keys.sort.each do |fact_name|
+      name_record = fact_name_class.new(fact_name_attributes(fact_name))
+
+      name_record = fact_name_class.where(name: fact_name).first unless name_record.save
```

Comment:
1. consider `find_by(name: fact_name)`
2. why not do it the other way around? e.g. check if the FactName exists. If not, create it. Is this the attempt to prevent the race condition? I don't get, why you create the record first in memory, save it and if that fails you try to load it from the database. If this is a check, if the record exists, I think it's not ideal. What happens if saving fails for another reason then that the record already exists?

Reply:
> Is this the attempt to prevent the race condition?

Yes, it is. I have already loaded everything that was in the DB on the first attempt (in the `initialize_fact_names` function). Here I am dealing with new fact names, so optimistically they all should be saved successfully. In the (rare?) case of race condition, I want to reuse the record from the DB.

> What happens if saving fails for another reason then that the record already exists?

You are right, I'll add `validate!` to ensure the record could be saved before the actual save attempt. Then after reloading, I could check that the record exists and throw if it doesn't - it will eliminate other failures.